File: neetcode150/graphs/leetcode133.py
from collections import defaultdict, deque


class Node:
    def __init__(self, val=0, neighbors=None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []


class Solution:
    def cloneGraph(self, node: "Node") -> "Node":
        """
        Clone Graph

        Approach:
            * We should keep copies of the node within the hashmap probably?

        :param node:
        :return:
        """
        if not node:
            return node

        graph = defaultdict()

        visited = set()

        queue = deque()
        queue.append(node)

        while queue:
            front = queue.popleft()

            if front.val in visited:
                continue

            visited.add(front.val)

            if front.val not in graph:
                graph[front.val] = Node(front.val)

            for neighbor in front.neighbors:
                if neighbor.val not in graph:
                    graph[neighbor.val] = Node(neighbor.val)

                graph[front.val].neighbors.append(graph[neighbor.val])

                if neighbor.val not in visited:
                    queue.append(neighbor)

        return graph[node.val]

File: js_algorithms/neetcode150/graphs/leetcode133.js
function Node(val, neighbors) {
  this.val = val ?? 0;
  this.neighbors = neighbors ?? [];
}

/**
 * @param {Node} node
 * @return {Node}
 */
const cloneGraph = function (node) {
  if (!node) {
    return null;
  }

  // go through each node in the graph and
  // create copies as we iterate through the graph
  // we should keep track of nodes we've visited via a set
  const copies = new Map();
  const result = new Node(node.val);
  const queue = [];

  queue.push([result, node]);
  copies.set(node.val, result);

  while (queue.length) {
    const [copy, orig] = queue.shift();

    for (const neighbor of orig.neighbors) {
      if (copies.has(neighbor.val)) {
        copy.neighbors.push(copies.get(neighbor.val));
      } else {
        const neighborCopy = new Node(neighbor.val);
        copy.neighbors.push(neighborCopy);
        copies.set(neighbor.val, neighborCopy);
        queue.push([neighborCopy, neighbor]);
      }
    }
  }

  return result;
};

const n1 = new Node(1);
const n2 = new Node(2);
const n3 = new Node(3);
const n4 = new Node(4);

n1.neighbors.push(n2);
n1.neighbors.push(n4);
n2.neighbors.push(n3);
n2.neighbors.push(n1);
n3.neighbors.push(n2);
n3.neighbors.push(n4);
n4.neighbors.push(n3);
n4.neighbors.push(n1);

console.log(cloneGraph(n1));
